fem/kernels/geometry.cpp

```diff
@@ -0,0 +1,302 @@
+Geometry* Geometry::Get(const FiniteElementSpace& fes,
+                        const IntegrationRule& ir,
+                        const Vector& Sx)
+{
+   const Mesh *mesh = fes.GetMesh();
+   const GridFunction *nodes = mesh->GetNodes();
+   const FiniteElementSpace *fespace = nodes->FESpace();
+   const FiniteElement *fe = fespace->GetFE(0);
+   const int dims     = fe->GetDim();
+   const int numDofs  = fe->GetDof();
+   const int numQuad  = ir.GetNPoints();
+   const int elements = fespace->GetNE();
+   const int ndofs    = fespace->GetNDofs();
+   const kDofQuadMaps* maps = kDofQuadMaps::GetSimplexMaps(*fe, ir);
+   NodeCopyByVDim(elements,numDofs,ndofs,dims,geom->eMap,Sx,geom->meshNodes);
+   Geom(dims, numDofs, numQuad, elements,
+        maps->dofToQuadD,
+        geom->meshNodes, geom->J, geom->invJ, geom->detJ);
+   delete maps;
+   return geom;
+}
+
+// *****************************************************************************
+Geometry* Geometry::Get(const FiniteElementSpace& fes,
+                        const IntegrationRule& ir)
+{
+   Mesh *mesh = fes.GetMesh();
+   const bool geom_to_allocate = !geom;
+   if (geom_to_allocate) { geom = new Geometry(); }
+   if (!mesh->GetNodes())
+   {
+      // mesh->SetCurvature(1, false, -1, Ordering::byVDIM);
+   }
+   const GridFunction *nodes = mesh->GetNodes();
+
+   const mfem::FiniteElementSpace *fespace = nodes->FESpace();
+   const mfem::FiniteElement *fe = fespace->GetFE(0);
+   const int dims     = fe->GetDim();
+   const int elements = fespace->GetNE();
+   const int numDofs  = fe->GetDof();
+   const int numQuad  = ir.GetNPoints();
+   const bool orderedByNODES = (fespace->GetOrdering() == Ordering::byNODES);
+   if (orderedByNODES) { ReorderByVDim(nodes); }
+   const int asize = dims*numDofs*elements;
+   mfem::Array<double> meshNodes(asize);
+   const Table& e2dTable = fespace->GetElementToDofTable();
+   const int* elementMap = e2dTable.GetJ();
+   mfem::Array<int> eMap(numDofs*elements);
+   GeomFill(dims,
+            elements,
+            numDofs,
+            elementMap,
+            eMap.GetData(),
+            nodes->GetData(),
+            meshNodes.GetData());
+
+   if (geom_to_allocate)
+   {
+      geom->meshNodes.allocate(dims, numDofs, elements);
+      geom->eMap.allocate(numDofs, elements);
+   }
+   kernels::vector::Assign(asize, meshNodes.GetData(), geom->meshNodes);
+   ArrayAssign(numDofs*elements, eMap.GetData(), geom->eMap);
+   // Reorder the original gf back
+   if (orderedByNODES) { ReorderByNodes(nodes); }
+   if (geom_to_allocate)
+   {
+      geom->J.allocate(dims, dims, numQuad, elements);
+      geom->invJ.allocate(dims, dims, numQuad, elements);
+      geom->detJ.allocate(numQuad, elements);
+   }
+   const kDofQuadMaps* maps = kDofQuadMaps::GetSimplexMaps(*fe, ir);
+   Geom(dims, numDofs, numQuad, elements, maps->dofToQuadD,
+        geom->meshNodes, geom->J, geom->invJ, geom->detJ);
```

Comment:
Are J, invJ, and detJ always allocated? This is a lot of extra memory used if this is only assembled once, or if the mesh moves. Do we still have the option (from the OCCA port) to provide a mask?